File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/search/RadioSearchEngineListPreference.java
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.search;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.RadioGroup;

import org.mozilla.focus.R;
import org.mozilla.focus.utils.Settings;

public class RadioSearchEngineListPreference extends SearchEngineListPreference implements RadioGroup.OnCheckedChangeListener {

    public RadioSearchEngineListPreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public RadioSearchEngineListPreference(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected View onCreateView(ViewGroup parent) {
        final View view = super.onCreateView(parent);
        searchEngineGroup.setOnCheckedChangeListener(this);
        return view;
    }

    @Override
    protected int getItemResId() {
        return R.layout.search_engine_radio_button;
    }

    @Override
    protected void updateDefaultItem(CompoundButton defaultButton) {
        defaultButton.setChecked(true);
    }


    @Override
    public void onCheckedChanged (RadioGroup group, int checkedId) {
        Settings.getInstance(group.getContext()).setDefaultSearchEngine(searchEngines.get(checkedId));
    }
}

File: mobile/android/focus-android/app/src/main/java/org/mozilla/focus/search/RadioSearchEngineListPreference.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.mozilla.focus.search

import android.content.Context
import android.util.AttributeSet
import android.widget.CompoundButton
import android.widget.RadioGroup
import androidx.preference.PreferenceViewHolder
import mozilla.components.browser.state.search.SearchEngine
import org.mozilla.focus.GleanMetrics.SearchEngines
import org.mozilla.focus.R
import org.mozilla.focus.ext.components

private const val ENGINE_TYPE_CUSTOM = "custom"
private const val ENGINE_TYPE_BUNDLED = "bundled"

class RadioSearchEngineListPreference : SearchEngineListPreference, RadioGroup.OnCheckedChangeListener {

    override val itemResId: Int
        get() = R.layout.search_engine_radio_button

    @Suppress("unused")
    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    @Suppress("unused")
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun onBindViewHolder(holder: PreferenceViewHolder) {
        super.onBindViewHolder(holder)
        searchEngineGroup!!.setOnCheckedChangeListener(this)
    }

    override fun updateDefaultItem(defaultButton: CompoundButton) {
        defaultButton.isChecked = true
    }

    override fun onCheckedChanged(group: RadioGroup, checkedId: Int) {
        val selectedEngine = group.getChildAt(checkedId) ?: return

        // check if the corresponding button was pressed or a11y focused.
        val hasProperState = selectedEngine.isPressed || selectedEngine.isAccessibilityFocused

        /* onCheckedChanged is called intermittently before the search engine table is full, so we
           must check these conditions to prevent crashes and inconsistent states. */
        if (group.childCount != searchEngines.count() || !hasProperState) {
            return
        }

        val newDefaultEngine = searchEngines[checkedId]

        context.components.searchUseCases.selectSearchEngine(newDefaultEngine)

        val source = if (newDefaultEngine.type == SearchEngine.Type.CUSTOM) {
            ENGINE_TYPE_CUSTOM
        } else {
            ENGINE_TYPE_BUNDLED
        }

        SearchEngines.setDefault.record(SearchEngines.SetDefaultExtra(source))
    }
}
